File: python/rings-and-rods.py
class Solution:
    def countPoints(self, rings: str) -> int:
        color = 'E'
        colors_in_rod = defaultdict(lambda: 0)
        for c in rings:
            if color == 'R':
                colors_in_rod[c] |= 1
            if color == 'G':
                colors_in_rod[c] |= 2
            if color == 'B':
                colors_in_rod[c] |= 4
            color = c
        point_count = 0
        for rod in colors_in_rod:
            if colors_in_rod[rod] == 7:
                point_count += 1
        return point_count

File: rust/rings-and-rods.rs
use std::collections::HashMap;

impl Solution {
    pub fn count_points(rings: String) -> i32 {
        let mut colors_in_rod = HashMap::new();
        let mut color = 'E';
        for c in rings.chars() {
            match color {
                'R' => *colors_in_rod.entry(c).or_insert(0) |= 1,
                'G' => *colors_in_rod.entry(c).or_insert(0) |= 2,
                'B' => *colors_in_rod.entry(c).or_insert(0) |= 4,
                _  => color = c
            }
            color = c;
        }
        let mut points = 0;
        for (key, colors) in colors_in_rod {
            if colors == 7 {
                points += 1;
            }
        }
        return points;
    }
}
